### PR TITLE
Allow multiple and other authors on articles

### PR DESCRIPTION
**Pull request by jonasdeluna**
# Description

Option to add multiple/other users as author on articles.

# Result

Before:

image

After:

image

Before:

image

After:

image

After:

image

# Testing

- [ X] I have thoroughly tested my changes.

---

Depends on: https://github.com/webkom/lego/pull/3220
Resolves: https://github.com/webkom/lego/issues/3195

### CODE REVIEW

File: app/components/Form/ObjectPermissions.tsx [hunk 1]
```diff
@@ -24,12 +24,14 @@ const ObjectPermissions = ({
...
   canViewGroups,
   requireAuth,
+  author,
   ...props
 }: {
   canEditUsers?: any;
   canEditGroups?: any;
   canViewGroups?: any;
   requireAuth?: any;
+  author?: any;
```
Comment (LudvigHz) [inline app/components/Form/ObjectPermissions.tsx:34]: Would be nice if you could set the correct types here :)

File: app/components/Form/ObjectPermissions.tsx [hunk 2]
```diff
@@ -100,12 +113,19 @@ export const normalizeObjectPermissions = ({
           canViewGroups,
         }
       : {}),
+
+    ...(author
+      ? {
+          author,
+        }
+      : {}),
```
Comment (LudvigHz) [inline app/components/Form/ObjectPermissions.tsx:121]: It would be awesome if you could rewrite this god-aweful code as well!

File: app/components/Form/ObjectPermissions.tsx [hunk 3]
```diff
@@ -122,6 +142,11 @@ export const objectPermissionsToInitialValues = ({
     initialCanEditUsers
       .filter(Boolean)
       .map((user) => ({ ...user, label: user.fullName, value: user.id }));
+  const author =
+    initialAuthors &&
+    initialAuthors
```
Comment (LudvigHz) [inline app/components/Form/ObjectPermissions.tsx:147]: `initialAuthors?.filter(...`

File: app/routes/articles/ArticleEditRoute.ts [hunk 4]
```diff
@@ -19,13 +20,24 @@ const mapStateToProps = (state, props) => {
   const article = selectArticleById(state, {
     articleId,
   });
+  const author = article?.author?.map((e) => {
+    return selectUserById(state, {
+      userId: e,
+    });
+  });
+  console.log(article);
```
Comment (LudvigHz) [inline app/routes/articles/ArticleEditRoute.ts:28]: Remove this

File: app/routes/articles/ArticleEditRoute.ts [hunk 5]
```diff
@@ -19,13 +20,24 @@ const mapStateToProps = (state, props) => {
...
     articleId,
     isNew: false,
     initialValues: {
       ...article,
-      ...objectPermissionsToInitialValues(article),
+      ...objectPermissionsToInitialValues({
+        canViewGroups: article.canViewGroups,
+        canEditGroups: article.canEditGroups,
+        canEditUsers: article.canEditUsers,
+        author: author,
```
Comment (LudvigHz) [inline app/routes/articles/ArticleEditRoute.ts:39]: Why is this here? :thinking:

Comment (LudvigHz) [reply app/routes/articles/ArticleEditRoute.ts:39]: Ah, if you need to add `author` here, change the function name to make sense with this change

File: app/routes/articles/components/ArticleDetail.tsx [hunk 6]
```diff
@@ -70,16 +70,31 @@ const ArticleDetail = ({
...
-        <span className={styles.detail}>
-          Skrevet av
-          <Link to={`/users/${author.username}`}> {author.fullName}</Link>
-        </span>
-        <span className={styles.detail}>
-          {moment(article.createdAt).format('lll')}
-        </span>
-      </div>
-
+      {
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleDetail.tsx:73]: This is not needed here

File: app/routes/articles/components/ArticleDetail.tsx [hunk 7]
```diff
@@ -70,16 +70,31 @@ const ArticleDetail = ({
...
+        <div className={styles.articleDetails}>
+          <span className={styles.detail}>
+            Skrevet av{' '}
+            {author &&
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleDetail.tsx:77]: This is superfluous with the check below.

File: app/routes/articles/components/ArticleDetail.tsx [hunk 8]
```diff
@@ -70,16 +70,31 @@ const ArticleDetail = ({
...
+                return (
+                  <span key={e.username}>
+                    <Link
+                      to={`/users/${e.username}`}
+                      className={styles.overviewAuthor}
+                    >
+                      {' '}
+                      {e.fullName}
+                    </Link>
+                    {e === author[author.length - 1] ? '' : ','}
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleDetail.tsx:88]: I think using the index directly is easier here

File: app/routes/articles/components/ArticleEditor.tsx [hunk 9]
```diff
@@ -192,12 +193,12 @@ const onSubmit = (
     ...normalizeObjectPermissions(data),
     youtubeUrl: data.youtubeUrl,
     title: data.title,
-    author: currentUser.id,
     description: data.description,
     content: data.content,
     tags: (data.tags || []).map((tag) => tag.value.toLowerCase()),
     pinned: data.pinned,
   };
+  body.author = body.author?.length !== 0 ? body.author : [currentUser.id];
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleEditor.tsx:201]: Inline this in the initialization instead of modifying it outside here.

File: app/routes/articles/components/Overview.tsx [hunk 10]
```diff
@@ -24,17 +24,21 @@ const OverviewItem = ({ article }: { article: ArticleWithAuthorDetails }) => (
...
-            to={`/users/${article.author.username}`}
-            className={styles.overviewAuthor}
-          >
-            {' '}
-            {article.author.fullName}
-          </Link>{' '}
-        </span>
-      )}
+      {article?.author &&
+        article?.author?.map((e) => {
```
Comment (LudvigHz) [inline app/routes/articles/components/Overview.tsx:28]: Here as well, when you use the `?.` operator, you don't need the `article?.author &&` check.

**Review by LudvigHz — CHANGES_REQUESTED** [review]
Cool!

You should rename `author` to `authors` to reflect this change. And fix up the small other things and this is good to go :)

Regarding removing the created_by, I think it might be a nice thing to keep?

**Comment by jonasdeluna** [comment]
On default the author field will be the current user, and if the field is empty it will also be set to the current user upon validation as a security measure.

File: app/routes/articles/ArticleDetailRoute.ts [hunk 11]
```diff
@@ -56,7 +58,7 @@ export default compose(
   ),
   connect(mapStateToProps, mapDispatchToProps),
   loadingIndicator(['article.content']),
-  helmet((props: { article: PublicArticle; author: PublicUser }, config) => {
+  helmet((props: { article: PublicArticle; authors: PublicUser }, config) => {
```
Comment (LudvigHz) [inline app/routes/articles/ArticleDetailRoute.ts:61]: authors is an array now :) so you need to change this to `PublicUser[]`

File: app/routes/articles/ArticleDetailRoute.ts [hunk 12]
```diff
@@ -104,8 +106,8 @@ export default compose(
         content: props.article.description,
       },
       {
-        property: 'article:author',
-        content: `${config.webUrl}/users/${props.author.username}`,
+        property: 'article:authors',
+        content: `${config.webUrl}/users/${props.authors.username}`,
```
Comment (LudvigHz) [inline app/routes/articles/ArticleDetailRoute.ts:110]: Ref. that authors is now an array, this will fail.

See https://ogp.me/#type_article for the correct way to handle multiple authors in the meta tags here

File: app/routes/articles/ArticleListRoute.ts [hunk 13]
```diff
@@ -34,8 +34,10 @@ const mapStateToProps = (state, props) => {
     }
   ).map((article) => ({
     ...article,
-    author: selectUserById(state, {
-      userId: article.author,
+    authors: article?.authors?.map((e) => {
```
Comment (LudvigHz) [inline app/routes/articles/ArticleListRoute.ts:37]: `article` should not be nullish here?

File: app/routes/articles/components/ArticleDetail.tsx (hunk 6 above)
Comment (LudvigHz) [reply app/routes/articles/components/ArticleDetail.tsx:73]: This is still here :grimacing:

File: app/routes/articles/components/Overview.tsx [hunk 14]
```diff
@@ -24,17 +24,17 @@ const OverviewItem = ({ article }: { article: ArticleWithAuthorDetails }) => (
...
-          <Link
-            to={`/users/${article.author.username}`}
-            className={styles.overviewAuthor}
-          >
-            {' '}
-            {article.author.fullName}
-          </Link>{' '}
-        </span>
-      )}
+      {article?.authors?.map((e) => {
```
Comment (LudvigHz) [inline app/routes/articles/components/Overview.tsx:27]: Here as well, `article` should not suddenly be nullish?

Comment (LudvigHz) [reply app/routes/articles/components/Overview.tsx:27]: Realisticly, `authors` is also a required field and should always be defined as well no?

File: app/components/Form/ObjectPermissions.tsx [hunk 15]
```diff
@@ -24,12 +25,14 @@ const ObjectPermissions = ({
...
 }: {
-  canEditUsers?: any;
-  canEditGroups?: any;
-  canViewGroups?: any;
-  requireAuth?: any;
+  canEditUsers?: PublicGroup[];
+  canEditGroups?: PublicGroup[];
+  canViewGroups?: PublicGroup[];
+  requireAuth?: boolean;
+  authors?: number[];
```
Comment (LudvigHz) [inline app/components/Form/ObjectPermissions.tsx:35]: Hmm, I realize now that this is all in ObjectPermission related stuff.

I don't know how fan I am of this, since authors don't really have anything to do with ObjectPermissions and is specific to just articles.

File: app/routes/articles/components/ArticleEditor.tsx [hunk 16]
```diff
@@ -189,15 +190,15 @@ const onSubmit = (
           cover: data.cover,
         }
       : {}),
-    ...normalizeObjectPermissions(data),
+    ...normalizeObjectPermissions({ ...data, currentUser: currentUser }),
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleEditor.tsx:193]: You shouldn't do this. It is more than enough that you set the default to the current user. If they remove themselves from the authors field, but it sets them as the author anyways that would be very strange.

**Review by LudvigHz — CHANGES_REQUESTED** [review]
Sorry for suddenly finding more stuff here, I should have done a more thorough review the first time around :disappointed:.

A couple things (as are mentioned in the comments):
- If authors is a required field in the backend, it should always be defined here, and both the types, and the usage of the `authors` field should reflect that. So no null checks or anything here (array length checks are fine). Let the type system catch it.
- Right now, the `authors` field is set together with the `ObjectPermissions`. This is not really related, and ObjectPermissions are generic fields that can be used on any field that use objectPermissions. I do like the idea of having default fields that can be reused super easy, but I doubt that we would have authors anywhere else than articles so with that in mind, both that it's not generic and that it does not have anything to do with ObjectPermissions, I think you should move it out and put the code for authors in the articleeditor directly.

File: app/routes/articles/components/ArticleEditor.tsx (hunk 16 above)
Comment (ivarnakken) [reply app/routes/articles/components/ArticleEditor.tsx:193]: Is this not just to ensure there's always an author? If they remove themselves, and no author has been specified, it should default to `currentUser`, no? However, if they remove themselves and add someone else, it should just be that other user. I feel like that's the "better" MMI.

Comment (jonasdeluna) [reply app/routes/articles/components/ArticleEditor.tsx]: Made the field required with front-end validation. An empty array seems to be allowed in the back-end but unless someone explicitly sends a request this should not happen.

` authors: data.authors.length > 0 ? data.authors.map((e) => e.value) : null,`

File: app/routes/articles/ArticleEditRoute.ts [hunk 17]
```diff
@@ -19,13 +21,26 @@ const mapStateToProps = (state, props) => {
...
       ...article,
-      ...objectPermissionsToInitialValues(article),
+      ...objectToInitialValues({
+        canViewGroups: article.canViewGroups,
+        canEditGroups: article.canEditGroups,
+        canEditUsers: article.canEditUsers,
+      }),
+      authors: authors
+        .filter(Boolean)
+        .map((user) => ({ ...user, label: user.fullName, value: user.id })),
```
Comment (LudvigHz) [inline app/routes/articles/ArticleEditRoute.ts:43]: Is there a reason you spread `...user` here?

File: app/routes/articles/ArticleListRoute.ts [hunk 18]
```diff
@@ -34,8 +34,10 @@ const mapStateToProps = (state, props) => {
     }
   ).map((article) => ({
     ...article,
-    author: selectUserById(state, {
-      userId: article.author,
+    authors: article.authors.map((e) => {
+      return selectUserById(state, {
+        userId: e,
+      });
```
Comment (LudvigHz) [inline app/routes/articles/ArticleListRoute.ts:40]: You should _ideally_ use a selector for multiple users here. Not critical though

File: app/routes/articles/components/ArticleEditor.tsx [hunk 19]
```diff
@@ -196,24 +213,43 @@ const onSubmit = (
           cover: data.cover,
         }
       : {}),
-    ...normalizeObjectPermissions(data),
+    ...normalizeObjectPermissions({ ...data }),
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleEditor.tsx:216]: Why the spread here?

File: app/routes/articles/components/ArticleEditor.tsx [hunk 20]
```diff
@@ -196,24 +213,43 @@ const onSubmit = (
...
     tags: (data.tags || []).map((tag) => tag.value.toLowerCase()),
     pinned: data.pinned,
   };
+
   return submitArticle(body);
 };
 
+type ValidationError<T> = Partial<{
+  [key in keyof T]: string | Record<string, string>[];
+}>;
```
Comment (LudvigHz) [inline app/routes/articles/components/ArticleEditor.tsx:231]: Loving the types😍

**Review by LudvigHz — APPROVED** [review]
Nice 👍 looks good now. Up to you if you want to fix up the last issues

File: app/routes/articles/components/ArticleEditor.tsx [hunk 21]
```diff
@@ -196,24 +213,43 @@ const onSubmit = (
...
+    data.youtubeUrl
+  );
+
+  if (!isValidYoutubeUrl) {
+    errors.youtubeUrl = errorMessage;
+  }
+
+  if (!data.authors || data.authors.length === 0) {
+    errors.title = 'Forfatter er påkrevd';
+  }
```
[the remark this answers is not in the document]
Comment (ivarnakken) [reply app/routes/articles/components/ArticleEditor.tsx:245]: This is entirely fine imo. though.

**Comment by ivarnakken** [comment]
Any reason to not merge this in now? @jonasdeluna You've done great work!

**Review by ivarnakken — APPROVED** [review]

**Comment by jonasdeluna** [comment]
> Any reason to not merge this in now? @jonasdeluna You've done great work!

Thanks, not really unless the validator needs changing ill have a final look at the test^

**Comment by jonasdeluna** [comment]
Everything should be good, but messed up the rebase :| also the backend tests are failing...
